File: 1b/node/src/frequencer.ts
export class Frequencer {
  intSequence: number[];
  maxAttempts: number;

  constructor(strSequence: string[]) {
    this.intSequence = strSequence.map(x => Number(x));
    this.maxAttempts = 999999;
  }

  sum() {
    return this.intSequence.reduce(this._reduceSum, 0);
  }

  findResonance() {
    const length = this.intSequence.length;
    let cursor = 0;
    let total = 0;
    let results = [0];

    while (cursor < this.maxAttempts) {
      let index = cursor % length;
      let num = this.intSequence[index];
      total += num;

      if (results.indexOf(total) !== -1) {
        return total;
      }

      results.push(total);
      cursor++;
    }

    throw new Error(
      "Unable to find a number to satisfy the criteria before hitting maximum attempts"
    );
  }

  _reduceSum(total: number, num: number) {
    return total + num;
  }
}

File: 1b/python/puzzle_1b/frequencer.py
class Frequencer:

    def __init__(self, str_sequence):
        self.int_sequence = list(map(int, str_sequence))
        self.max_attempts = 999999

    def sum(self):
        return sum(self.int_sequence)

    def find_resonance(self):
        cursor = 0
        total = 0
        sums = set([0])

        while cursor < self.max_attempts:
            index = cursor % len(self.int_sequence)
            number = self.int_sequence[index]
            total += number

            if total in sums:
                return total

            sums.add(total)
            cursor += 1

        raise OverflowError(
            'exceeded maximum number of attempts: {}'.format(self.max_attempts))
